- Type

        1. Integer
        2. Float
        3. Complex

- Operation

        1. numbers              — Numeric abstract base classes
        2. math                 — Mathematical functions
        3. cmath                — Mathematical functions for complex numbers
        4. decimal              — Decimal fixed point and floating point arithmetic
        5. fractions            — Rational numbers
        6. random               — Generate pseudo-random numbers
        7. statistics           — Mathematical statistics functions

# <center> Numeric types</center> 

In [ ]:
#Input
b2=int(input("Enter int  :  "))  

## 1. int( )

In [233]:
#help(int)

In [234]:
b=int()
type(b)

int

In [235]:
# Calling function
val = int(10) # integer value
val2 = int(10.52) # float value
val3 = int('10') # string value
# Displaying result
print("integer values :",val, val2, val3) 

integer values : 10 10 10


In [236]:
b1=10
b1

10

In [237]:
type(b1)

int

## c. float( )

In [ ]:
#help(float)

In [ ]:
c=float()
type(c)

In [ ]:
c1=10.5
print(c1)
c1

In [ ]:
type(c1)

In [ ]:
c2 = float(input("Enter FLOAT  :  "))
c2

## d. complex( )

In [2]:
#help(complex)

In [ ]:
d=complex()
type(d)

In [ ]:
# Python complex() function example
# Calling function
a = complex(1) # Passing single parameter
b = complex(1,2) # Passing both parameters
# Displaying result
print(a)
print(b) 

In [ ]:
d1=2.0 + 2.3j
print(d1)
d1

In [ ]:
#Complex
x = 3+5j
y = 5j
z = -5j

In [ ]:
type(d1)

In [ ]:
d2 = complex(input("Enter COMPILE  :  "))     #2+2.3j
d2

# <center>Type Casting</center>


1. int()     - Integer
2. float()   - Float Value
3. chr()     - Charecter Value
4. complex

In [ ]:
#Integers:
int(1)   # x will be 1

In [ ]:
int(2.8) # y will be 2

In [ ]:
int("3") # z will be 3

In [ ]:
#Floats:
float(1)     # x will be 1.0

In [ ]:
float(2.8)   # y will be 2.8

In [ ]:
float("3")   # z will be 3.0

In [ ]:
float("4.2") # w will be 4.2

In [ ]:
#Strings:
str("s1") # x will be 's1'

In [ ]:
str(2)    # y will be '2'

In [ ]:
str(3.0)  # z will be '3.0'

**Note :**
    
 - can't convert complex to int 
 - invalid literal for int() with base 10:

# <center>Built-in Functions</center>


In [ ]:
#help(__builtins__)

## 3. Math fuction

**abs()**

In [ ]:
# help(abs)

In [64]:
print('Absolute value of -40 is:', abs(-20))

Absolute value of -40 is: 20


In [66]:
print('Absolute value of -40.83 is:', abs(-20.83))

Absolute value of -40.83 is: 20.83


**sum()**

In [ ]:
#help(sum)

In [62]:
sum([1, 2,4 ])

7

In [61]:
sum([1, 2, 4], 10)

17

**round()**

In [ ]:
#help(round)

In [58]:
round(10)  

10

In [59]:
round(10.8)

11

In [60]:
round(6.6)

7

**pow() / Root**

In [ ]:
#help(pow)
# positive x, positive y (x**y)  

In [67]:
pow(4, 2)

16

In [68]:
pow(-4, 2)

16

In [69]:
pow(4, -2)

0.0625

In [70]:
pow(-4, -2)

0.0625

**divmod()**

In [ ]:
#help(divmod)

In [74]:
divmod(11,2)

(5, 1)

In [75]:
divmod(11,2)[0]

5

In [76]:
divmod(11,2)[1]

1

**eval()**

In [ ]:
#help(eval)

In [77]:
x = 8
eval('x + 1')

9

# <center>Other Lib</center>

In [ ]:
https://mathworld.wolfram.com/topics/CombinatorialGeometry.html
    
https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.jamesbrennan.org%2Falgebra%2Fnumbers%2Freal_number_system.htm&psig=AOvVaw0K2qi1v15kWLI96oGJ4lHI&ust=1596723933259000&source=images&cd=vfe&ved=0CAIQjRxqFwoTCPDSraqihOsCFQAAAAAdAAAAABAO

## 1. numbers

- A complex number is a number that can be expressed in the form a + bi, where a and b are real numbers, and i satisfies the equation i2 = −1. Because no real number satisfies this equation, i is called an imaginary number. For the complex number a + bi, a is called the real part, and b is called the imaginary part.

- Number
    - Complex
        - Real
            - Rational
                 - Integral

In [ ]:
import numbers
help(numbers)

## 2. math

In [1]:
from math import *

In [8]:
print("pi   :",pi)        #3.141592653589793
print("e    :",e)          #2.718281828459045
print("inf  :",inf)   #inf
print("nan  :",nan)      #nan
print("tau  :",tau)      #6.283185307179586

pi   : 3.141592653589793
e    : 2.718281828459045
inf  : inf
nan  : nan
tau  : 6.283185307179586


In [ ]:
# isfinite(x, /)  : Return True if x is neither an infinity nor a NaN, and False otherwise.
# isinf(x, /)     : Return True if x is a positive or negative infinity, and False otherwise.
# isnan(x, /)     : Return True if x is a NaN (not a number), and False otherwise.

**Basic**

In [ ]:
#floor(x, /):Return the floor of x as an Integral.
#This is the largest integer <= x.

In [ ]:
#fmod(x, y, /):Return fmod(x, y), according to platform C.
#x % y may differ.

In [ ]:
#exp(x, /):Return e raised to the power of x.

In [ ]:
#sqrt(x, /): Return the square root of x.

In [ ]:
#modf(x, /) : Return the fractional and integer parts of x.
        
        Both results carry the sign of x and are floats.

In [ ]:
#factorial(x, /):Find x!.
factorial(5)

### Trignomatric

In [25]:
radList=[
0,0.017452,0.034899,0.052336,0.069756,0.087156,0.104528,0.121869,0.139173,0.156434,
0.173648,0.190809,0.207912,0.224951,0.241922,0.258819,0.275637,0.292372,0.309017,0.325568,
 0.34202,0.358368,0.374607,0.390731,0.406737,0.422618,0.438371,0.45399,0.469472,0.48481,
0.5,0.515038,0.529919,0.544639,0.559193,0.573576,0.587785,0.601815,0.615661,0.62932,
0.642788,0.656059,0.669131,0.681998,0.694658,0.707107,0.71934,0.731354,0.743145,0.75471,
0.766044,0.777146,0.788011,0.798636,0.809017,0.819152,0.829038,0.838671,0.848048,0.857167,
0.866025,0.87462,0.882948,0.891007,0.898794,0.906308,0.913545,0.920505,0.927184,0.93358,
0.939693,0.945519,0.951057,0.956305,0.961262,0.965926,0.970296,0.97437,0.978148,0.981627,
0.984808,0.987688,0.990268,0.992546,0.994522,0.996195,0.997564,0.99863,0.999391,0.999848,1]

In [26]:
degList=list(range(0,90,1))

In [24]:
deg_rad_chat={deg:rad for deg, rad in zip(degList,radList)}

def sinfun(deg=None,rad=None):
    if deg:
        
        
    elif rad:
    
    else 
    
    
sinfun(20)

π rad = 180°

1Rad × 180/π = 57.296Deg

α(degrees) = α(radians) × 180° / π

degrees = radians × 180° / π

<table class="ntable">
<tbody><tr>
<th>Radians (rad)</th>
<th>Radians (rad)</th>
<th>Degrees (°)</th>
</tr>
<tr>
<td>0 rad</td>
<td>0 rad</td>
<td>0°</td>
</tr>
<tr>
<td>π/6 rad</td>
<td>0.5235987756 rad</td>
<td>30°</td>
</tr>
<tr>
<td>π/4 rad</td>
<td>0.7853981634 rad</td>
<td>45°</td>
</tr>
<tr>
<td>π/3 rad</td>
<td>1.0471975512 rad</td>
<td>60°</td>
</tr>
<tr>
<td>π/2 rad</td>
<td>1.5707963268 rad</td>
<td>90°</td>
</tr>
<tr>
<td>2π/3 rad</td>
<td>2.0943951024 rad</td>
<td>120°</td>
</tr>
<tr>
<td>3π/4 rad</td>
<td>2.3561944902 rad</td>
<td>135°</td>
</tr>
<tr>
<td>5π/6 rad</td>
<td>2.6179938780 rad</td>
<td>150°</td>
</tr>
<tr>
<td>π rad</td>
<td>3.1415926536 rad</td>
<td>180°</td>
</tr>
<tr>
<td>3π/2 rad</td>
<td>4.7123889804 rad</td>
<td>270°</td>
</tr>
<tr>
<td>2π rad</td>
<td>6.2831853072 rad</td>
<td>360°</td>
</tr>
</tbody></table>

#degrees(x, /):Convert angle x from radians to degrees.
degrees(1.5707963267948966)

#radians(x, /):Convert angle x from degrees to radians.
radians(90)

**rad**

#sin(x, /) : Return the sine of x (measured in radians).   
sin(1.5707963267948966)

sin(0)

#cos(x, /) : Return the cosine of x (measured in radians).
cos(0)

cos(1)

#tan(x, /) : Return the tangent of x (measured in radians).

#asin(x, /) : Return the arc sine (measured in radians) of x.  

#acos(x, /) : Return the arc cosine (measured in radians) of x.    

#atan(x, /) : Return the arc tangent (measured in radians) of x.

#atan2(y, x, /) : Return the arc tangent (measured in radians) of y/x.
        
        Unlike atan(y/x), the signs of both x and y are considered.
    

**hyperbolic**

In [9]:
#sinh(x, /) : Return the hyperbolic sine of x.

#cosh(x, /) : Return the hyperbolic cosine of x.

#tanh(x, /):Return the hyperbolic tangent of x.

#asinh(x, /) : Return the inverse hyperbolic sine of x.

#acosh(x, /) : Return the inverse hyperbolic cosine of x.

#atanh(x, /) : Return the inverse hyperbolic tangent of x.
    

### log(...)

 : log(x, [base=math.e])
        Return the logarithm of x to the given base.
        
        If the base not specified, returns the natural logarithm (base e) of x.
    
    log10(x, /)
        Return the base 10 logarithm of x.
    
    log1p(x, /)
        Return the natural logarithm of 1+x (base e).
        
        The result is computed in a way which is accurate for x near zero.
    
    log2(x, /)
        Return the base 2 logarithm of x.
    

### Function

In [ ]:
#ceil(x, /):Return the ceiling of x as an Integral.
#his is the smallest integer >= x.
    

copysign(x, y, /):Return a float with the magnitude (absolute value) of x but the sign of y.

    On platforms that support signed zeros, copysign(1.0, -0.0)
    returns -1.0.


#lgamma(x, /):Natural logarithm of absolute value of Gamma function at x.

#erf(x, /):Error function at x.
#erfc(x, /):Complementary error function at x.
#expm1(x, /):Return exp(x)-1.
This function avoids the loss of precision involved in the direct evaluation of exp(x)-1 for small x.
    

#fabs(x, /):Return the absolute value of the float x.
   

frexp(x, /):Return the mantissa and exponent of x, as pair (m, e).
        m is a float and e is an int, such that x = m * 2.**e.
        If x is 0, m and e are both 0.  Else 0.5 <= abs(m) < 1.0.
    
    fsum(seq, /):Return an accurate floating point sum of values in the iterable seq.
        
        Assumes IEEE-754 floating point arithmetic.
    
    gamma(x, /):Gamma function at x.
    
    gcd(x, y, /):greatest common divisor of x and y
    
    hypot(x, y, /)
        Return the Euclidean distance, sqrt(x*x + y*y).
    
    isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)
        Determine whether two floating point numbers are close in value.
        
          rel_tol
            maximum difference for being considered "close", relative to the
            magnitude of the input values
          abs_tol
            maximum difference for being considered "close", regardless of the
            magnitude of the input values
        
        Return True if a is close in value to b, and False otherwise.
        
        For the values to be considered close, the difference between them
        must be smaller than at least one of the tolerances.
        
        -inf, inf and NaN behave similarly to the IEEE 754 Standard.  That
        is, NaN is not close to anything, even itself.  inf and -inf are
        only close to themselves.

    ldexp(x, i, /)
        Return x * (2**i).
        
        This is essentially the inverse of frexp().
    
        pow(x, y, /)
        Return x**y (x to the power of y).
        

    

    

    

    
    remainder(x, y, /)
        Difference between x and the closest integer multiple of y.
        
        Return x - n*y where n*y is the closest integer multiple of y.
        In the case where x is exactly halfway between two multiples of
        y, the nearest even value of n is used. The result is always exact.



    

In [ ]:
    trunc(x, /)
        Truncates the Real x to the nearest Integral toward 0.
        
        Uses the __trunc__ magic method.

## 3. cmath

## 4. decimal

## 5. fractions

In [ ]:
from fractions import Fraction

Fraction(16, -10)

Fraction(123)

Fraction()

Fraction('3/7')

Fraction('1.414213')

Fraction('7e-6')

## 6. random

## 7. statistics